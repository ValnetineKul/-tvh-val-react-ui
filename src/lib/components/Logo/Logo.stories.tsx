import React from 'react';
import { Link } from 'react-router-dom';
import Logo from './Logo';
import { createMeta } from '../../story-utils';

type Props = React.ComponentProps<typeof Logo>;
const Template = (args: Props) => <Logo {...args} />;

export const Primitive = Template.bind({});
Primitive.args = {
  variant: 'positive',
  alt: 'Logo alternative text',
  size: 'sm',
  component: Link,
  to: '/whereToGoTo',
  bepcoVariant: 'company',
};
Primitive.storyName = 'Logo';

export default createMeta({
  component: Logo,
  title: 'Foundations/Logo',
  argTypes: {
    onClick: {
      action: { argTypesRegex: '^on.*' },
      control: { disable: true },
    },
    component: {
      control: {
        disable: true,
      },
      table: {
        type: { summary: 'React.ReactElement' },
      },
    },
    href: {
      control: {
        disable: true,
      },
      table: {
        type: { summary: 'If "href" is filled, the component will immediately transform to HTML link tag "a"' },
      },
    },
    disabled: {
      table: {
        disable: true,
      },
    },
    download: {
      control: {
        disable: true,
      },
    },
  },
});
