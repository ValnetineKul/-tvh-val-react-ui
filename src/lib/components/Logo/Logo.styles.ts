import { makeStyles } from '../../../themes/core';

const useStyles = makeStyles({ name: 'Logo' })((theme) => ({
  root: {
    '&:not(:disabled):hover': {
      backgroundColor: 'transparent',
    },
    '&:not(:disabled):active': {
      backgroundColor: 'transparent',
    },
  },
  sizeSm: {
    height: 24,
  },
  sizeMd: {
    height: 40,
  },
  withSubLogo: {
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'start',
    rowGap: theme.layoutSpacing['spacing/100'],
  },
  sizeCompanySm: {
    height: 14,
  },
  sizeCompanyMd: {
    height: 24,
  },
  sizeSubLogoSm: {
    height: 8,
  },
  sizeSubLogoMd: {
    height: 14,
  },
}));

export default useStyles;
