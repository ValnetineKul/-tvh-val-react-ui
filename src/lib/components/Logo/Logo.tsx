import React from 'react';
import type { FC, MouseEventHandler } from 'react';
import { capitalize } from '@mui/material/utils';
import { useTheme, useSurface } from '../../../themes/core';
import ButtonBase from '../Surfaces/ButtonBase';
import type { ButtonBaseProps } from '../ButtonBase';
import { useButtonBase } from '../ButtonBase';
import Image from '../MediaItems/Image';
import useStyles from './Logo.styles';

export interface LogoProps extends ButtonBaseProps {
  alt?: string;
  subAlt?: string;
  size?: 'sm' | 'md';
  variant?: 'positive' | 'negative';
  bepcoVariant?: 'company' | 'productBrand';
  onClick?: MouseEventHandler<HTMLButtonElement>;
  className?: string;
}

const Logo: FC<LogoProps> = ({
  alt = 'Logo',
  subAlt = 'A company of TVH.',
  size = 'sm',
  variant = 'positive',
  bepcoVariant = 'company',
  onClick,
  className,
  ...props
}) => {
  const { color } = useSurface();
  const { classes, cx } = useStyles();

  const buttonBaseProps = useButtonBase(props);

  const sizeClassName = (classes as Record<string, string>)[`size${capitalize(size)}`];
  const sizeCompanyClassName = (classes as Record<string, string>)[`sizeCompany${capitalize(size)}`];
  const sizeSubLogoClassName = (classes as Record<string, string>)[`sizeSubLogo${capitalize(size)}`];

  const theme = useTheme();

  const withCompanySubLogo = bepcoVariant === 'productBrand' && theme.logo.subLogo;

  const renderLogo = () => {
    if (withCompanySubLogo) {
      return (
        <span className={classes.withSubLogo}>
          <Image src={theme.logo[variant]} alt={alt} fallback="text" className={sizeCompanyClassName} />
          <Image src={theme.logo.subLogo} alt={subAlt} fallback="text" className={sizeSubLogoClassName} />
        </span>
      );
    }
    return <Image src={theme.logo[variant]} alt={alt} fallback="text" className={sizeClassName} />;
  };

  return (
    <ButtonBase color={color} className={cx(classes.root, className)} onClick={onClick} {...buttonBaseProps}>
      {renderLogo()}
    </ButtonBase>
  );
};

export default Logo;
