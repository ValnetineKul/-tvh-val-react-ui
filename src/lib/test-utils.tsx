import userEvent from '@testing-library/user-event';
import { fireEvent, render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { theme } from '../themes/tvh';
import { createTheme, ThemeProvider, SurfaceProvider } from '../themes/core';

export const muiTheme = createTheme(theme);

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <MemoryRouter>
      <ThemeProvider theme={muiTheme}>
        <SurfaceProvider color="100">{children}</SurfaceProvider>
      </ThemeProvider>
    </MemoryRouter>
  );
};

const customRender = (component: React.ReactElement, options?: Parameters<typeof render>[1]) =>
  render(component, { wrapper: AllTheProviders, ...options });

const customRenderHook: typeof renderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

export const initResponsiveTest = (size: number | 'Mobile' | 'Tablet' | 'Desktop' | 'LgDesktop' | 'XlDesktop') => {
  let width: number;
  if (size === 'Mobile') width = muiTheme.breakpoints.values.sm - 1;
  else if (size === 'Tablet') width = muiTheme.breakpoints.values.sm;
  else if (size === 'Desktop') width = muiTheme.breakpoints.values.md;
  else if (size === 'LgDesktop') width = muiTheme.breakpoints.values.lg;
  else if (size === 'XlDesktop') width = muiTheme.breakpoints.values.xl;
  else width = size;

  Object.defineProperty(window, 'innerWidth', {
    writable: true,
    configurable: true,
    value: width,
  });

  window.matchMedia = jest.fn().mockImplementation((query) => {
    const queryWidthInterval: string[] = query.match(/\d+(\.\d+)?/g);

    let matches;
    if (queryWidthInterval.length === 2) {
      matches = width >= +queryWidthInterval[0] && width < +queryWidthInterval[1];
    } else {
      const queryWidth = +queryWidthInterval[0];
      const isDown = /max/.test(query);

      matches = isDown ? width <= queryWidth : width >= queryWidth;
    }

    return {
      matches,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  });
};

const customUserEvent = Object.assign(userEvent, {
  mouseDown: fireEvent.mouseDown,
  mouseUp: fireEvent.mouseUp,
  mouseMove: fireEvent.mouseMove,
  scroll: fireEvent.scroll,
  error: fireEvent.error,
});

export { screen, within, waitFor, act, waitForElementToBeRemoved } from '@testing-library/react';
// override render method
export { customRender as render, customRenderHook as renderHook, customUserEvent as userEvent };
