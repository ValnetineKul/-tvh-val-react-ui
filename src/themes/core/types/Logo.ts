import type React from 'react';

export type Logo = {
  positive: React.ReactNode;
  negative?: React.ReactNode;
};
