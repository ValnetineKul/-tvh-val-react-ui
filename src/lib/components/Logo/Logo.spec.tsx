import React from 'react';
import type { ComponentProps } from 'react';
import { render, screen } from '../../test-utils';
import Logo from './Logo';

type Props = ComponentProps<typeof Logo>;

describe('Logo', () => {
  it('Should render correctly', () => {
    const { baseElement } = render(<Logo size="sm" />);
    expect(baseElement).toMatchSnapshot();
  });

  describe('Should set logo size', () => {
    const cases: [Props['size'], string][] = [
      [undefined, 'Logo-sizeSm'],
      ['sm', 'Logo-sizeSm'],
      ['md', 'Logo-sizeMd'],
    ];
    test.each(cases)('given %p, returns %p', (input, expected) => {
      render(<Logo size={input} />);
      const image = screen.getByAltText('Logo');
      expect(image.className).toMatch(expected);
    });
  });

  it('Should set custom alt text', () => {
    const customAlt = 'Alt text';
    render(<Logo alt={customAlt} />);
    const logoImage = screen.getByAltText(customAlt);
    expect(logoImage).toBeInTheDocument();
  });
});
