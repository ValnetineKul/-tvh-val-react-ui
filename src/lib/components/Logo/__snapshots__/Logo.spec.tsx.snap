// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logo Should render correctly 1`] = `
<body>
  <div>
    <button
      class="MuiButtonBase-root tss-1xk2c6w-ButtonBase-root-ButtonBase-focusOut-Logo-root mui-10d1a0h-MuiButtonBase-root"
      tabindex="0"
      type="button"
    >
      <img
        alt="Logo"
        class="tss-iftyg6-Image-root-Logo-sizeSm"
        src="positiveLogo.svg"
      />
    </button>
  </div>
</body>
`;
